File: tools/python/xen/xend/server/event.py
#============================================================================
# This library is free software; you can redistribute it and/or
# modify it under the terms of version 2.1 of the GNU Lesser General Public
# License as published by the Free Software Foundation.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#============================================================================
# Copyright (C) 2004, 2005 Mike Wray <mike.wray@hp.com>
# Copyright (C) 2005 XenSource Ltd
#============================================================================

import sys
import StringIO

from xen.web import protocol, tcp, unix

from xen.xend import scheduler
from xen.xend import sxp
from xen.xend import PrettyPrint
from xen.xend.XendError import XendError
from xen.xend import XendLogging
from xen.xend import XendRoot


xroot = XendRoot.instance()


DEBUG = 0

class EventProtocol(protocol.Protocol):
    """Asynchronous handler for a connected event socket.
    """

    def __init__(self, daemon):
        #protocol.Protocol.__init__(self)
        self.daemon = daemon
        # Event queue.
        self.queue = []
        self.parser = sxp.Parser()
        self.pretty = 1

    def dataReceived(self, data):
        try:
            self.parser.input(data)
            while(self.parser.ready()):
                val = self.parser.get_val()
                res = self.dispatch(val)
                self.send_result(res)
            if self.parser.at_eof():
                self.loseConnection()
        except SystemExit:
            raise
        except:
            self.send_error()

    def loseConnection(self):
        if self.transport:
            self.transport.loseConnection()
        if self.connected:
            scheduler.now(self.connectionLost)

    def connectionLost(self, reason=None):
        pass

    def send_reply(self, sxpr):
        io = StringIO.StringIO()
        if self.pretty:
            PrettyPrint.prettyprint(sxpr, out=io)
        else:
            sxp.show(sxpr, out=io)
        print >> io
        io.seek(0)
        if self.transport:
            return self.transport.write(io.getvalue())
        else:
            return 0

    def send_result(self, res):
        if res is None:
            resp = ['ok']
        else:
            resp = ['ok', res]
        return self.send_reply(resp)

    def send_error(self):
        (extype, exval) = sys.exc_info()[:2]
        return self.send_reply(['err',
                                ['type', str(extype)],
                                ['value', str(exval)]])

    def send_event(self, val):
        return self.send_reply(['event', val[0], val[1]])

    def queue_event(self, name, v):
        # Despite the name we don't queue the event here.
        # We send it because the transport will queue it.
        self.send_event([name, v])
        
    def opname(self, name):
         return 'op_' + name.replace('.', '_')

    def operror(self, name, req):
        raise XendError('Invalid operation: ' +name)

    def dispatch(self, req):
        op_name = sxp.name(req)
        op_method_name = self.opname(op_name)
        op_method = getattr(self, op_method_name, self.operror)
        return op_method(op_name, req)

    def op_help(self, _1, _2):
        def nameop(x):
            if x.startswith('op_'):
                return x[3:].replace('_', '.')
            else:
                return x
        
        l = [ nameop(k) for k in dir(self) if k.startswith('op_') ]
        return l

    def op_quit(self, _1, _2):
        self.loseConnection()

    def op_exit(self, _1, _2):
        sys.exit(0)

    def op_pretty(self, _1, _2):
        self.pretty = 1

    def op_info(self, _1, _2):
        val = ['info']
        #val += self.daemon.blkifs()
        #val += self.daemon.netifs()
        #val += self.daemon.usbifs()
        return val

    def op_trace(self, _, v):
        mode = (v[1] == 'on')
        self.daemon.tracing(mode)

    def op_log_stderr(self, _, v):
        mode = v[1]
        if mode == 'on':
            XendLogging.addLogStderr()
        else:
            XendLogging.removeLogStderr()

    def op_domain_ls(self, _1, _2):
        xd = xroot.get_component("xen.xend.XendDomain")
        return xd.list_names()

    def op_domain_configure(self, _, v):
        domid = sxp.child_value(v, "dom")
        config = sxp.child_value(v, "config")
        if domid is None:
            raise XendError("missing domain id")
        if config is None:
            raise XendError("missing domain config")
        xd = xroot.get_component("xen.xend.XendDomain")
        xd.domain_configure(domid, config)

    def op_domain_unpause(self, _, v):
        domid = sxp.child_value(v, "dom")
        if domid is None:
            raise XendError("missing domain id")
        xd = xroot.get_component("xen.xend.XendDomain")
        xd.domain_unpause(domid)

class EventFactory(protocol.ServerFactory):
    """Asynchronous handler for the event server socket.
    """

    def __init__(self, daemon):
        protocol.ServerFactory.__init__(self)
        self.daemon = daemon

    def buildProtocol(self, _):
        return EventProtocol(self.daemon)

def listenEvent(daemon):
    factory = EventFactory(daemon)
    if xroot.get_xend_unix_server():
        path = '/var/lib/xend/event-socket'
        unix.listenUNIX(path, factory)
    if xroot.get_xend_http_server():
        port = xroot.get_xend_event_port()
        interface = xroot.get_xend_address()
        l = tcp.listenTCP(port, factory, interface=interface)
        l.setCloExec()

File: tools/ocaml/xenstored/event.ml
(*
 * Copyright (C) 2006-2007 XenSource Ltd.
 * Copyright (C) 2008      Citrix Ltd.
 * Author Vincent Hanquez <vincent.hanquez@eu.citrix.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation; version 2.1 only. with the special
 * exception on linking described in file LICENSE.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *)

(**************** high level binding ****************)
type t = {
  handle: Xeneventchn.handle;
  domexc: Xeneventchn.t;
}

(* On clean start, both parameters will be None, and we must open the evtchn
   handle and bind the DOM_EXC VIRQ.  On Live Update, the fd is preserved
   across exec(), and the DOM_EXC VIRQ still bound. *)
let init ?fd ?domexc_port () =
  let handle = match fd with
    | None -> Xeneventchn.init ~cloexec:false ()
    | Some fd -> fd |> Utils.FD.of_int |> Xeneventchn.fdopen
  in
  let domexc = match domexc_port with
    | None -> Xeneventchn.bind_dom_exc_virq handle
    | Some p -> Xeneventchn.of_int p
  in
  { handle; domexc }

let fd eventchn = Xeneventchn.fd eventchn.handle
let bind_interdomain eventchn domid port = Xeneventchn.bind_interdomain eventchn.handle domid port
let unbind eventchn port = Xeneventchn.unbind eventchn.handle port
let notify eventchn port = Xeneventchn.notify eventchn.handle port
let pending eventchn = Xeneventchn.pending eventchn.handle
let unmask eventchn port = Xeneventchn.unmask eventchn.handle port

let dump e chan =
  Printf.fprintf chan "evtchn-dev,%d,%d\n"
    (Utils.FD.to_int @@ Xeneventchn.fd e.handle)
    (Xeneventchn.to_int e.domexc)
